### PR TITLE
In-memory Engine: encode internal key trailer in little endian.

### LINKED ISSUE #17082: In-memory Engine: u64 endian in key encoding is wrong
## Bug Report

https://github.com/tikv/tikv/blob/290dab1617c138e790bed26b7c127de2a58adfde/components/region_cache_memory_engine/src/keys.rs#L182-L198

The comments state that the `encode_internal_bytes` matches rocksdb internal key encoding, but actually it's not.
It encodes `(seq << 8) | v_type` in big endian which is the oppopsite to rocksdb's little endian.
The actually bytes representation is 
```
user key (n bytes) | sequence number shifted by 1 bytes (7 bytes) | value type (1 byte)
```

### What version of TiKV are you using?

Nightly 2024-05-29

### Steps to reproduce

N/A

### What did you expect?

It matches rocksdb's encoding.

### PR DESCRIPTION
**Pull request by overvenus**
### What is changed and how it works?

Issue Number: Close #17082

What's Changed:

```commit-message
An internal key trailer is `(sequence_number << 8) | value_type as u64`
and it should be encoded in little endian to be consistent with RocksDB.
```

### Check List

Tests 

- Unit test

### Release note

```release-note
None
```

### CODE REVIEW

**Review by SpadeA-Tang — APPROVED** [review]

File: components/region_cache_memory_engine/src/keys.rs [hunk 1]
```diff
@@ -180,9 +180,9 @@ pub fn decode_key(encoded_key: &[u8]) -> InternalKey<'_> {
 }
 
 /// Format for an internal key (used by the skip list.)
-/// ```
-/// contents:      key of size n     | value type | sequence number shifted by 8 bits
-/// byte position:         0 ..  n-1 | n          |  n + 1 .. n + 7
+///
+/// ```text
+/// Format: | user key (n bytes) | value type (1 bytes) | sequence number (7 bytes) |
```
Comment (zhangjinpeng87) [inline components/region_cache_memory_engine/src/keys.rs:185]: According to https://github.com/facebook/rocksdb/blob/v9.2.1/db/dbformat.h#L178 , the format should be 
`user key | seq number (7 bytes) | value type (1 byte)`

Comment (zhangjinpeng87) [reply components/region_cache_memory_engine/src/keys.rs:185]: emm, seems after encode by little endian, the format is `user key | value type (1 byte) | seq number (7 bytes)`.

File: components/region_cache_memory_engine/src/keys.rs [hunk 2]
```diff
@@ -309,4 +311,17 @@ mod tests {
...
+    #[test]
+    fn test_encode_decode() {
+        let encoded_bytes = encode_internal_bytes(b"foo", 7, ValueType::Value);
+        let key = decode_key(encoded_bytes.as_slice());
+        assert_eq!(key.sequence, 7);
+        assert_eq!(key.v_type, ValueType::Value as _);
+
+        let value_type_byte = encoded_bytes.as_slice()[encoded_bytes.as_slice().len() - 8];
+        assert_eq!(value_type_byte, ValueType::Value as u8);
+        let seq_lsb = encoded_bytes.as_slice()[encoded_bytes.as_slice().len() - 7];
```
Comment (zhangjinpeng87) [inline components/region_cache_memory_engine/src/keys.rs:324]: should use `from_le_bytes` to read the u64.

File: components/region_cache_memory_engine/src/keys.rs (hunk 1 above)
Comment (overvenus) [reply components/region_cache_memory_engine/src/keys.rs:185]: It's encoded in little endian, so in the bytes representation, "value type" should be ahead of "seq number".

**Review by LykxSassinator — APPROVED** [review]

**Pull request merged**